codegen-server/src/test/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGeneratorTest.kt

```diff
@@ -54,4 +123,195 @@ class ServerBuilderGeneratorTest {
+                    if (optional || (it.key == "DocumentNull" && !requiredTrait)) {
+                        rust(".is_none(), true")
+                    } else if (it.key == "DocumentNull") {
+                        rust(", aws_smithy_types::Document::Null")
+                    } else if (it.key == "DocumentString") {
+                        rust(", String::from(${it.value}).into()")
+                    } else if (it.key.startsWith("DocumentNumber")) {
+                        val type = it.key.replace("DocumentNumber", "")
+                        rust(", aws_smithy_types::Document::Number(aws_smithy_types::Number::$type(${it.value}))")
+                    } else if (it.key.startsWith("Document")) {
+                        rust(", ${it.value}.into()")
+                    } else {
+                        rust(", " + it.value)
+                    }
+                    rust(");")
+                }
+        }
+    }
+
+    private fun generateModel(required: Boolean, values: Map<String, String?>): Model {
+        val requiredTrait = if (required) "@required" else ""
+
+        val members = values.entries.joinToString(", ") {
+            """
+            $requiredTrait
+            ${it.key.toPascalCase()}: ${it.key} = ${if (it.value == null) "null" else it.value}
+            """
+        }
+        val model =
+            """
+            namespace com.test
+            use smithy.framework#ValidationException
+
+            structure MyStruct {
+                $members
+            }
+
+            enum Language {
+                EN = "en",
+                FR = "fr",
+            }
+
+            list StringList {
+                member: String
+            }
+
+            map IntegerMap {
+                key: String
+                value: Integer
+            }
+
+            document DocumentNull
+            document DocumentBoolean
+            document DocumentString
+            document DocumentDecimal
+            document DocumentNumberNegInt
+            document DocumentNumberPosInt
+            document DocumentNumberFloat
+            document DocumentList
+            document DocumentMap
+            """
+        return model.asSmithyModel(smithyVersion = "2")
+    }
+
+    private fun localParameters(): Stream<Arguments> {
+        val builderWriters = listOf(::writeServerBuilderGenerator, ::writeServerBuilderGeneratorWithoutPublicConstrainedTypes)
+        return Stream.of(
+            TestConfig(customValues, requiredTrait = true, nullDefault = true, optional = false),
+            TestConfig(defaultValues, requiredTrait = false, nullDefault = false, optional = false),
+            TestConfig(customValues, requiredTrait = true, nullDefault = false, optional = false),
+            TestConfig(defaultValues, requiredTrait = false, nullDefault = true, optional = true),
+        ).flatMap { builderWriters.stream().map { builderWriter -> Arguments.of(it, builderWriter) } }
+    }
+
+    data class TestConfig(
```

Comment:
This is the most important piece of information a reader needs to understand what the test is testing. Each field needs documentation on how it impacts the generated code and the assertions we make.